lokbot/farmer.py

```diff
@@ -821,6 +827,73 @@ def academy_farmer_thread(self, to_max_level=False):
         threading.Timer(2 * 3600, self.academy_farmer_thread, [to_max_level]).start()
         return
 
+    def _troop_training_capacity(self):
+        """
+        return total troop training capacity of all barracks
+        """
+        buildings = self.kingdom_enter.get('kingdom', {}).get('buildings', [])
+        troop_training_capacity = 0
+        for building in buildings:
+            if building.get('state') == BUILDING_STATE_NORMAL:
+                continue
+
+            if building['code'] == BUILDING_CODE_MAP['barrack']:
+                troop_training_capacity += BARRACK_LEVEL_TROOP_TRAINING_RATE_MAP[int(building['level'])]
+
+        return troop_training_capacity
+
+    def _total_troops_capacity_according_to_resources(self, troop_code):
+        """
+        return maximum number of troops according to resources
+        """
+        req_resources = TRAIN_TROOP_RESOURCE_REQUIREMENT[troop_code]
+
+        amount = 0
+        for req_resource, resource in (req_resources, self.resources):
+            if amount == 0 or resource // req_resource <= amount:
+                amount = resource // req_resource
+
+        return amount
+
+    def train_troop_thread(self, troop_code):
+        current_tasks = self.api.kingdom_task_all().get('kingdomTasks', [])
+
+        worker_used = [t for t in current_tasks if t.get('code') == TASK_CODE_CAMP]
+
+        troop_training_capacity = self._troop_training_capacity()
+
+        if worker_used:
+            if worker_used[0].get('status') == STATUS_CLAIMED:
+                self.api.kingdom_task_claim(random.choice([105, 110]))
```

Comment:
Maybe we need to random choice the position code based on buildings which typed `barrack`, instead of hard-coded 105-110 ?